[docs] SDK 50 guide for Sentry

brentvatne opened the pull request:
# Why

Ensure we have instructions for using `@sentry/react-native` with SDK 50.

# TODO

- [ ] Test the latest release after speaking with the Sentry team on Jan 10
- [ ] Update this page to indicate that `sentry-expo` is deprecated in SDK 50

brentvatne committed:
[docs] First pass at SDK 50 beta guide for Sentry

krystofwoldrich commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+<Step label="1.2">
+
+Go to the [Sentry API section](https://sentry.io/settings/account/api/auth-tokens/), and create an
+**auth token**. The token requires the scopes: `org:read`, `project:releases`, and
+`project:write`. Save them.
```
Sentry has new organization auth tokens that are preferred for source maps uploaded as they are not bound to individuals but to org. 
```suggestion
Go to the [DEVELOPER SETTINGS -> Auth Tokens](https://sentry.io/settings/) and create a new token. The token is automatically scoped for Source Map Upload and Release Creation. Save it.
```

krystofwoldrich commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -333,6 +338,143 @@ Sentry.init({
+```json app.json
+{
+  "expo": {
+    "plugins": ["sentry-expo", {
````
The new plugin import
```suggestion
    "plugins": ["@sentry/react-native/expo", {
```

krystofwoldrich commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+> **warning** Never commit secrets to your repository. Keep the Sentry auth token in a safe place, outside your repository.
+
+<ConfigReactNative>
+
+If you do not use [Continuous Native Generation (CNG)](https://docs.expo.dev/workflow/continuous-native-generation/), then you should use the [@sentry/wizard](https://docs.sentry.io/platforms/react-native/#install).
+
+</ConfigReactNative>
```
This is true, but users can also use it to set up apps with CNG.

https://github.com/getsentry/sentry-wizard/pull/505

quinlanj commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -333,6 +338,143 @@ Sentry.init({
+Configuring `@sentry/react-native` can be done through the config plugin in your [app config](/workflow/configuration/).  Add the following to your project's **app.json** or **app.config.js** file:
+
+{/* prettier-ignore */}
+```json app.json
+{
+  "expo": {
+    "plugins": ["sentry-expo", {
+      "organization": "sentry org slug, or use the `SENTRY_ORG` environment variable",
+      "project": "sentry project name, or use the `SENTRY_PROJECT` environment variable"
+    }]
+    /* @hide ... your existing configuration */ /* @end */
+  }
+}
+```
````
Should be nested array as per https://docs.expo.dev/config-plugins/introduction/#use-a-config-plugin
```suggestion
```json app.json
{
  "expo": {
    "plugins": [["@sentry/react-native", {
      "organization": "sentry org slug, or use the `SENTRY_ORG` environment variable",
      "project": "sentry project name, or use the `SENTRY_PROJECT` environment variable"
    }]
    /* @hide ... your existing configuration */ /* @end */
    ]
  }
}
```
```

Simek commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+Create a new release build of your app and verify that it uploads source maps correctly. You may want to add a button in your app to test that it is working and sourcemaps are wired up as expected, for example:
+
+```js
````
```suggestion
```jsx
```

Nit, for betted code highlight. Also setting the code block title might be nice, for clarity, and to allow easier copy.

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+### Installation
+
+In your project directory, run:
```
```suggestion
Run the following command in your project directory to install `@sentry/react-native`:
```

Let's spell out what's happening in the command (good from accessibility view)

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+If you are using EAS Build, you can set the `SENTRY_AUTH_TOKEN` environment variable by [creating a secret of the same name](/build-reference/variables/#using-secrets-in-environment-variables).
+
+> **warning** Never commit secrets to your repository. Keep the Sentry auth token in a safe place, outside your repository.
```
I feel, we should provide more context on what that safe place can be, may be can mention one or two examples here?

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+<ConfigReactNative>
+
+If you do not use [Continuous Native Generation (CNG)](https://docs.expo.dev/workflow/continuous-native-generation/), then you should use the [@sentry/wizard](https://docs.sentry.io/platforms/react-native/#install).
```
```suggestion
If you do not use [Continuous Native Generation (CNG)](https://docs.expo.dev/workflow/continuous-native-generation/), then you should use the [`@sentry/wizard`](https://docs.sentry.io/platforms/react-native/#install).
```

Since this is a library, let's use inline syntax highlighting here.

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -333,6 +338,143 @@ Sentry.init({
+```js
+import * as Sentry from "@sentry/react-native";
+
+// your App component here
````
```suggestion
// Your App component here
```

We start all comments with captialziation.

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -333,6 +338,143 @@ Sentry.init({
+```js
+import { Button } from 'react-native';
+
+// inside some component
````
```suggestion
// Inside some component
```

amandeepmittal commented on docs/pages/guides/using-sentry.mdx:
```diff
@@ -333,6 +338,143 @@ Sentry.init({
+
+## Usage with EAS Update
+
+In the current state of SDK 50 beta, EAS Update is not yet supported. We're working on making it easier than ever to integrate updates with Sentry, and we plan to release along with SDK 50 proper.
```
```suggestion
In the current state of SDK 50 beta, EAS Update is not yet supported. We're working on making it easier than ever to integrate updates with Sentry, and we plan to release it along with SDK 50.
```

brentvatne committed:
Apply suggestions from code review

Co-authored-by: Aman Mittal <amandeepmittal@live.com>
Co-authored-by: Bartosz Kaszubowski <gosimek@gmail.com>

brentvatne committed:
Apply suggestions from code review

Co-authored-by: Kryštof Woldřich <31292499+krystofwoldrich@users.noreply.github.com>

brentvatne committed:
Apply suggestions from code review

Co-authored-by: Kryštof Woldřich <31292499+krystofwoldrich@users.noreply.github.com>

brentvatne committed:
Update to latest

quinlanj committed:
[docs] add sentry instructions for eas update (#26266)

# Why

Add EAS Update instructions for Sentry integration in SDK 50


# Test Plan

Inspected manually with `yarn run dev`
<!--
Please describe how you tested this change and how a reviewer could
reproduce your test, especially if this PR does not include automated
tests! If possible, please also provide terminal output and/or
screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).

krystofwoldrich commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,199 @@ import { CODE } from '~/ui/components/Text';
+
+<ImageSpotlight alt="Sentry update tags" src="/static/images/guides/sentry-tags.png" />
+
+### Upload source maps
+
+After running `eas update`, upload the source maps to Sentry:
+
+```sh
+SENTRY_AUTH_TOKEN=[super-secret-token] node node_modules/@sentry/react-native/scripts/expo-upload-sourcemaps.js dist
+```
````
Would it make sense to move this above the `Tag configuration` section? I know it logically makes sense to first set tags, but I believe uploaded source maps are more important than the tags and I don't want users to overlook this.

krystofwoldrich commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,199 @@ import { CODE } from '~/ui/components/Text';
+After running `eas update`, upload the source maps to Sentry:
+
+```sh
+SENTRY_AUTH_TOKEN=[super-secret-token] node node_modules/@sentry/react-native/scripts/expo-upload-sourcemaps.js dist
````
Let's use the `npx` command so we are not bound to the script path in the future. 

Plus it's shorter.

```suggestion
SENTRY_AUTH_TOKEN=[super-secret-token] npx sentry-expo-upload-sourcemaps dist
```

brentvatne committed:
Update docs/pages/guides/using-sentry.mdx

Co-authored-by: Kryštof Woldřich <31292499+krystofwoldrich@users.noreply.github.com>

brentvatne commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,199 @@ import { CODE } from '~/ui/components/Text';
+```js
+import * as Updates from 'expo-updates';
+
+const { metadata, extra } = Updates.manifest;
````
cc @quinlanj

quinlanj replied:
How about this: https://github.com/expo/expo/pull/25712#discussion_r1449709860

Open to suggestions for this suggestion 😄

wschurman replied:
Should be able to disambiguate the subtype by checking for the presence of `metadata`: https://github.com/expo/expo/blob/main/apps/expo-go/src/menu/DevMenuTaskInfo.tsx#L28

Something like:
```typescript
const { metadata, extra } =
    'metadata' in manifest ? manifest : { metadata: undefined, extra: undefined };
```
or this might be cleaner:
```typescript
const metadata = 'metadata' in manifest ? manifest.metadata : undefined;
const extra = 'metadata' in manifest ? manifest.extra : undefined;
```

`metadata` in the manifest has type `object` and `extra` has type `ManifestExtra`.

quinlanj replied:
based on will's feedback: https://github.com/expo/expo/pull/25712#discussion_r1450825925

In this snippet, `updateGroup` is typed as unknown, so we do a runtime check for 'string' where necessary

brentvatne replied:
would this also preserve types?

```
const { manifest } = Updates;
const metadata = manifest?.metadata;
const extra = metadata ? manifest.extra : undefined;
const updateGroup = metadata?.updateGroup;
```

Simek replied:
Can't we use, or cast to the exported `Manifest` type?
* https://github.com/expo/expo/blob/ab91257ace9c5a8196eacf6f7e2391f659538c15/packages/expo-updates/src/Updates.types.ts#L3

Or we can discern which exact type it is, by checking the presence of i.e. `metadata` key, via `in` operator before casting, if needed. So:

```tsx
const typedManifest = 'metadata' in manifest ? (manifest as ExpoUpdatesManifest) : (manifest as EmbeddedManifest); 
```

brentvatne replied:
this seems like a good suggestion @Simek. up to @quinlanj / @wschurman

wschurman replied:
It's also possible that it is empty object, though it's somewhat hard to tell when that's the case. Looks like it's when the library is disabled, either explicitly or due to an initialization error. So the actual type of `Updates.manifest` is `ExpoUpdatesManifest | EmbeddedManifest | {}` (I'll file a task to change it to not use `Partial<Manifest>`).

brentvatne committed:
Simplify updates section

fobos531 commented:
@brentvatne Sidenote, is Sentry yet to publish an updated version of their SDK which includes support for SDK 50? I tried migrating away from sentry-expo in one of my side projects using the instructions in this PR, and it couldn't find the config plugin.

brentvatne commented:
> @brentvatne Sidenote, is Sentry yet to publish an updated version of their SDK which includes support for SDK 50? I tried migrating away from sentry-expo in one of my side projects using the instructions in this PR, and it couldn't find the config plugin.

hello! yes! give this version a try: `"@sentry/react-native": "5.16.0-alpha.4"` - it will work with the info in this guide. a stable release with these changes is coming very soon

quinlanj commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,212 @@ import { CODE } from '~/ui/components/Text';
+
+Add the following snippet in the global scope as early as possible in your application's lifecycle.
+
+```js
+import * as Updates from 'expo-updates';
+
+const { metadata, extra } = Updates.manifest;
````
```suggestion
```ts
import * as Updates from 'expo-updates';

// cast if using typescript
const { metadata, extra } = Updates.manifest as Record<string, any>;
```

quinlanj commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,212 @@ import { CODE } from '~/ui/components/Text';
+
+Add the following snippet in the global scope as early as possible in your application's lifecycle.
+
+```js
+import * as Updates from 'expo-updates';
+
+const { metadata, extra } = Updates.manifest;
+
+Sentry.init({
+  // dsn, release, dist, etc...
+});
+
+Sentry.configureScope(scope => {
+  scope.setTag('expo-update-group-id', metadata?.updateGroup);
+  scope.setTag('expo-update-id', Updates.updateId);
+  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
+  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
+  scope.setTag(
+    'expo-update-debug-url',
+    Updates.isEmbeddedLaunch
+      ? 'not applicable for embedded updates'
+      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
+  );
+});
````
```suggestion
```js
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  scope.setTag(
    "expo-update-debug-url",
    Updates.isEmbeddedLaunch
      ? "not applicable for embedded updates"
      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${updateGroup}`
  );
});
```

brentvatne replied:
> `${extra?.expoClient?.owner}`

will this possibly lead to `/accounts/undefined` ever? or is this owner field always actually present and this is just a typing quirk?

> `/updates/${updateGroup}`

it looks like `updateGroup` can be undefined, so this would lead to `/updates/undefined` if so

Simek replied:
Can't we just use or cast to the exported `Manifest` type?
* https://github.com/expo/expo/blob/ab91257ace9c5a8196eacf6f7e2391f659538c15/packages/expo-updates/src/Updates.types.ts#L3

quinlanj replied:
Amended suggestion: https://github.com/expo/expo/pull/25712#discussion_r1452783256

on the `owner` field -- the docs say that this should fall back to username but I tested this and it is indeed `undefined`. I've made a change where it falls back to `[account]`, and the website should handle this case. The edge case is if `slug` is undefined but `owner` isnt. But most manifests have `slug` so that field should be relatively safe.

quinlanj commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,212 @@ import { CODE } from '~/ui/components/Text';
+Add the following snippet in the global scope as early as possible in your application's lifecycle.
+
+```js
+import * as Updates from 'expo-updates';
+
+const { metadata, extra } = Updates.manifest;
+
+Sentry.init({
+  // dsn, release, dist, etc...
+});
+
+Sentry.configureScope(scope => {
+  scope.setTag('expo-update-group-id', metadata?.updateGroup);
+  scope.setTag('expo-update-id', Updates.updateId);
+  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
+  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
+  scope.setTag(
+    'expo-update-debug-url',
+    Updates.isEmbeddedLaunch
+      ? 'not applicable for embedded updates'
+      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
+  );
+});
````
```suggestion
import * as Updates from 'expo-updates';

const { metadata, extra } = Updates.manifest;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope(scope => {
  scope.setTag('expo-update-group-id', metadata?.updateGroup);
  scope.setTag('expo-update-id', Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
  scope.setTag(
    'expo-update-debug-url',
    Updates.isEmbeddedLaunch
      ? 'not applicable for embedded updates'
      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
  );
});
```
```suggestion
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;
  
Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  if (Updates.isEmbeddedLaunch) {
    scope.setTag(
      "expo-update-debug-url",
      "not applicable for embedded updates"
    );
  } else if (typeof updateGroup === "string") {
    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  }
});
```

quinlanj commented on docs/pages/guides/using-sentry.mdx:
````diff
@@ -16,20 +18,212 @@ import { CODE } from '~/ui/components/Text';
+Add the following snippet in the global scope as early as possible in your application's lifecycle.
+
+```js
+import * as Updates from 'expo-updates';
+
+const { metadata, extra } = Updates.manifest;
+
+Sentry.init({
+  // dsn, release, dist, etc...
+});
+
+Sentry.configureScope(scope => {
+  scope.setTag('expo-update-group-id', metadata?.updateGroup);
+  scope.setTag('expo-update-id', Updates.updateId);
+  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
+  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
+  scope.setTag(
+    'expo-update-debug-url',
+    Updates.isEmbeddedLaunch
+      ? 'not applicable for embedded updates'
+      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
+  );
+});
````
```suggestion
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  if (Updates.isEmbeddedLaunch) {
    scope.setTag(
      "expo-update-debug-url",
      "not applicable for embedded updates"
    );
  } else if (typeof updateGroup === "string") {
    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  }
});
```

brentvatne replied:
@quinlanj - what do you think about this for the `configureScope` section instead?

```js
Sentry.configureScope((scope) => {
  scope.setTag("expo-update-id", Updates.updateId);
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);

  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);

    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  } else if (Updates.isEmbeddedLaunch) {
    // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
    if (Updates.isEmbeddedLaunch) {
      scope.setTag(
        "expo-update-debug-url",
        "not applicable for embedded updates"
      );
    }
  }
});
```

quinlanj replied:
lgtm!

brentvatne committed:
Add deprecation message and update code block for metadata

brentvatne committed:
Fix copy/paste typo
